Docs: update plugins tutorials (#373)
grafana/plugin-tools: docusaurus/docs/create-a-plugin/develop-a-plugin/build-a-data-source-plugin.md

  - datasource
---

import CreatePlugin from '@shared/create-plugin-frontend.md';
import PluginAnatomy from '@shared/plugin-anatomy.md';

## Introduction

Grafana supports a wide range of [data sources](https://grafana.com/grafana/plugins/data-source-plugins/), including Prometheus, MySQL, and Datadog. In some cases, though, you already have an in-house metrics solution that you’d like to add to your Grafana dashboards. This tutorial teaches you to build a new data source plugin to query data.

In this tutorial, you'll:

> - Build a data source to visualize a sine wave

Reviewer: ```suggestion
- Build a data source to visualize a sine wave.
```

> - Construct queries using the query editor

Reviewer: ```suggestion
- Construct queries using the query editor.
```

> - Configure your data source using the config editor

Reviewer: ```suggestion
- Configure your data source using the config editor.
```

### Prerequisites

- Grafana v9.0 or later
- [LTS](https://nodejs.dev/en/about/releases/) version of Node.js

## Create a new plugin

<CreatePlugin pluginType="datasource" />

> To learn how to create a backend data source plugin, see [Build a data source backend plugin](./build-a-data-source-backend-plugin.md)

Reviewer: ```suggestion
To learn how to create a backend data source plugin, refer to [Build a data source backend plugin](./build-a-data-source-backend-plugin.md).
```

The `options` object contains the queries, or _targets_, that the user made, along with context information, like the current time interval. Use this information to query an external database.

### Test your data source

> `testDatasource` implements a health check for your data source. For example, Grafana calls this method whenever the user clicks the **Save & Test** button, after changing the connection settings.

Reviewer: ```suggestion
The `testDatasource` method implements a health check for your data source. For example, Grafana calls `testDatasource` whenever the user clicks the **Save & Test** button, after changing the connection settings.
```

async testDatasource()
```

> For an example of a health check in a frontend data source, see our [datasource-http](https://github.com/grafana/grafana-plugin-examples/blob/edf9f0259d28bc14aaaac4204058f7caab99d6ab/examples/datasource-http/src/DataSource.ts#L84) plugin.

Reviewer: ```suggestion
For an example of a health check in a frontend data source, refer to our [datasource-http](https://github.com/grafana/grafana-plugin-examples/blob/edf9f0259d28bc14aaaac4204058f7caab99d6ab/examples/datasource-http/src/DataSource.ts#L84) plugin.
```

> ## Returning Data frames

Reviewer: ```suggestion
## Returning data frames
```

There are countless different databases, each with their own ways of querying data. To be able to support all the different data formats, Grafana consolidates the data into a unified data structure called [data frames](../../introduction/data-frames.md).

Let's see how to create and return a data frame from the `query` method. In this step, you'll change the code in the starter plugin to return a [sine wave](https://en.wikipedia.org/wiki/Sine_wave).

1. In the current `query` method, remove the code inside the `map` function.

>    The `query` method now look like this:

Reviewer: ```suggestion
   The `query` method now looks like this:
```
